hasher-matcher-actioner/hmalib/lambdas/actions/action_performer.py

```diff
@@ -10,24 +12,122 @@
 
 def lambda_handler(event, context):
     """
-    This lambda is called once per match per dataset. If a single hash matches
-    multiple datasets, this will be called multiple times.
-
-    Eventually, this will be just an action labeller. It will label a match
-    record with the action it recommends. A separate system will be stood up
-    that aggregates labels and 'decides' which action would be taken.
-
-    For now, this method will,
-    - it will construct the MatchMessage object and identify the actions it
-      needs to invoke.
-    - for now, rather than fanning out to individual specific lambdas, it will
-      call all specific functions serially!
+    This lambda is called when one or more matches are found. If a single hash matches
+    multiple datasets, this will be called only once.
+
+    Action labels are generated for each match message, then an action is performed
+    corresponding to each action label.
     """
-    for sqs_record in event["Records"]:
+    for sqs_record in event[
+        "Records"
+    ]:  # TODO research max # sqs records / lambda_handler invocation
         sns_notification = json.loads(sqs_record["body"])
         match_message: MatchMessage = MatchMessage.from_sns_message(
             sns_notification["Message"]
         )
+        action_labels = get_action_labels(match_message)
+        for action_label in action_labels:
+            # TODO create a new action execution queue and enqueue the
+            # match message and action label (or, possibly, add the
+            # action label to the match message and enqueue the match
+            # message by itself)
+            perform_action(match_message, action_label)
+
+        if threat_exchange_reacting_is_enabled(match_message):
+            threat_exchange_reaction_labels = get_threat_exchange_reaction_labels(
+                match_message, action_labels
+            )
+            if threat_exchange_reaction_labels:
+                for threat_exchange_reaction_label in threat_exchange_reaction_labels:
+                    # TODO create a new ThreatExchange reaction queue and enqueue
+                    # the match message and threat exchange reaction label (or, possibly,
+                    # add the threat exchange reaction label to the match message
+                    # and enqueue the match message by itself)
+                    react_to_threat_exchange(
+                        match_message, threat_exchange_reaction_label
+                    )
```

Comment:
See comment above. I think lines 36-39 should go above line 29 and lines 40-48 should go in the perform_action fucntion (ie. I was planning on handling these in the action executor).

Is this some thing you considered and decided against? Should we discuss?

Reply:
see above.